src/validations/html/rules/MismatchedTag.js

```diff
@@ -0,0 +1,139 @@
+        locationKey, {name, mismatchedTags});
+
+    if (opened) {
+      this._openTagStack = this._openTagStack.slice(0, openIndex);
+    } else {
+      this._unopenedTagLocations = this._unopenedTagLocations.add(locationKey);
+      this._unopenedTagQueuesByName =
+        this._unopenedTagQueuesByName.update(
+          name, List.of(), tags => tags.push({name, location}));
+
+      // If no matching open tag exists, then assume later close tags could
+      // match some open tag instead of emptying openTagStack.
+    }
+  }
+
+  // If any spare unopened tag exists, diagnose them and only yield
+  // MISPLACED_CLOSE_TAG.
+  //
+  // If no spare exists:
+  //
+  // A mismatched close tag that has been opened means the case
+  //
+  // <div>
+  //   <span>
+  // </div>
+  //
+  // so yield UNCLOSED_TAG.
+  //
+  // A mismatched close tag that has yet to be opened means the case
+  //
+  // <div>
+  //   </span>
+  // </div>
+  //
+  // so yield UNOPENED_TAG.
+  //
+  // Note: MISPLACED_CLOSE_TAG suggestions invalidate UNCLOSED_TAG and
+  // UNOPENED_TAG suggestions by moving other mismatched close tags.
+  // Diagnosing UNCLOSED_TAG and UNOPENED_TAG accurately would require another
+  // pass after all the MISPLACED_CLOSE_TAG suggestions are resolved.
+  * done() {
+    yield* this._yieldMisplacedCloseTags();
+    if (this._foundMisplacedCloseTag) {
+      return;
+    }
+
+    // Process the remaining errors.
+    for (const [location, {name, mismatchedTags}]
+         of this._mismatchedTagErrorsByLocation) {
+      if (this._unopenedTagLocations.has(location)) {
+        yield {
+          code: Code.UNOPENED_TAG,
+          closeTag: {
+            name,
+            location: {
+              row: location.get('row'),
+              column: location.get('column'),
+            },
+          },
+        };
+      } else {
+        for (const mismatchedTag of mismatchedTags) {
+          yield {
+            code: Code.UNCLOSED_TAG,
+            openTag: mismatchedTag,
+            closeTag: {
+              name,
+              location: {
+                row: location.get('row'),
+                column: location.get('column'),
+              },
+            },
+          };
+        }
+      }
+    }
+  }
+
+  * _yieldMisplacedCloseTags() {
+    for (const [location, {name, mismatchedTags}]
+         of this._mismatchedTagErrorsByLocation) {
+      for (const mismatchedTag of mismatchedTags) {
+        const spareUnopenedTag =
+          this._unopenedTagQueuesByName.get(
+            mismatchedTag.name, List.of()).first();
+        if (spareUnopenedTag) {
+          this._foundMisplacedCloseTag = true;
+          this._unopenedTagQueuesByName = this._unopenedTagQueuesByName.update(
+            spareUnopenedTag.name, tags => tags.shift());
+
+          yield {
+            code: Code.MISPLACED_CLOSE_TAG,
+            openTag: mismatchedTag,
+            closeTag: {
+              name,
+              location: {
```

Comment:
I would make a slightly stronger claim that the location of this error should be the location of the `spareUnopenedTag` rather than the location of the mismatched closing tag, e.g.:

```html
<!DOCTYPE html>
<html>
    <head>
        <title>Page Title</title>
    </head>
    <body>
        <div>
            <p>
        </div>
        </p> <!-- fundamentally feels like the error is on this line -->
    </body>
</html>
```

I would change the message to e.g. `This closing </p> tag should go before the </div> on line 9`